File: Algorithms/Search_Algorithms/Fibonacci_Search/fibonacci_search.js
//JavaScript program for Fibonacci Search 

 /* 
  * Find an element in the given sorted array with the help of Fibonacci series
    in O(log N) time complexity. 
  * Let the length of given array be n [0...n-1] and the element to be searched be x.
  * Find the smallest Fibonacci number greater than or equal to n.
  * source: https://iq.opengenus.org/fibonacci-search/
  */

  //length/size: n , element to search: x
  function fibonacciSearch(n, arr, x){
    //Let this number be fb(M) [m’th Fibonacci number].
    //Let the two Fibonacci numbers preceding it be fb(M-1) [(m-1)’th Fibonacci number] 
    let fbMm2 = 0;
    let fbMm1 = 1;
    let fbM = fbMm1 + fbMm2;
    let offset = -1;

    //Find the smallest Fibonacci number greater than or equal to n
    while(fbM < n){
        fbMm2 = fbMm1;
        fbMm1 = fbM;
        fbM = fbMm1 + fbMm2;
    }


    //While the array has elements to be checked:
    //-> Compare x with the last element of the range covered by fb(M-2)
    //-> If x matches, return index value
    //-> Else if x is less than the element, move the third Fibonacci variable two Fibonacci down,

     let i = 0;

     while(fbM > 1){
        i = Math.min(offset + fbMm2, n - 1);

        if(arr[i] < x){
            fbM  = fbMm1;
            fbMm1 = fbMm2;
            fbMm2 = fbM - fbMm1;
            offset = i;
        } else if(arr[i] > x){
            fbM  = fbMm2;
            fbMm1 = fbMm1 - fbMm2;
            fbMm2 = fbM - fbMm1;
        } else
            return i;
     }

    //comparing the last element
    if(fbMm1 && arr[offset + 1] == x)
        return offset + 1;

    return -1;
}

let arr = [10, 15, 30, 60, 90, 100, 120];
let n = 7;
let x = 92;
console.log(fibonacciSearch(n, arr, x)); 

File: Algorithms/Searching-Algorithm/Fibonacci-Search/fibonacciSearch.py
arr = [10, 22, 30, 44, 56, 58, 60, 72, 105, 110, 160]
x = 60


def FibonacciGenerator(n):
    if n < 1:
        return 0
    elif n == 1:
        return 1
    else:
        return FibonacciGenerator(n - 1) + FibonacciGenerator(n - 2)


def FibonacciSearch(arr, x):
    m = 0
    while FibonacciGenerator(m) < len(arr):
        m = m + 1
    offset = -1
    while (FibonacciGenerator(m) > 1):
        i = min(offset + FibonacciGenerator(m - 2), len(arr) - 1)
        print('Current Element : ', arr[i])
        if (x > arr[i]):
            m = m - 1
            offset = i
        elif (x < arr[i]):
            m = m - 2
        else:
            return i
    if(FibonacciGenerator(m - 1) and arr[offset + 1] == x):
        return offset + 1
    return -1


print('Number found at index : ', FibonacciSearch(arr, x))
